### PR TITLE
Encapsulate spec definitions with a class

### PR DESCRIPTION
**Pull request by tgoodlet**
Allows for easier introspection of spec definitions including function
signatures and hook options. Originally introduced to address ~#47~ #15 and #43 which
requires keeping track of spec default arguments values.

### CODE REVIEW

**Comment by tgoodlet** [comment]
@nicoddemus hah nice catch. 
Yeah it's actually #15 and the accompanying PR #43.
I'll change the commit msg.

**Comment by tgoodlet** [comment]
@RonnyPfannschmidt yeah if you find anything let me know.

This isn't really all that important if you don't like it.
I just thought I'd grab the handy stuff from #43 which might make somone's life easier down the road.

**Comment by RonnyPfannschmidt** [comment]
well, i fail to see it atm, lets go ahead with it

**Comment by tgoodlet** [comment]
@RonnyPfannschmidt we can always leave this sitting until we need it. 
It's really only useful if we plan to introspect `hookspec` func signatures in the future.

**Comment by tgoodlet** [comment]
@fschulze take a look at this if you don't mind so we can figure out if it's going to meet your needs.
With this you can introspect spec args (and other things) using `pm.hook.<hookname>.spec`.

**Comment by fschulze** [comment]
@tgoodlet 
> @fschulze take a look at this if you don't mind so we can figure out if it's going to meet your needs.
With this you can introspect spec args (and other things) using pm.hook.<hookname>.spec.

That again is from the host side, do you have anything in mind for the plugin side? Is that even possible?

I still think my proposal like this would be simpler:

Host side:

```
@hookspec(deprecated=True)
def myhost_oldhook(...):
    ...

@hookspec(replaces=myhost_oldhook)
def myhost_newhook(...):
    ...
```

Plugin side:
```
@hookimpl(optional=True)
def myhost_oldhook(...):
    ...

@hookimpl(optional=True)
def myhost_newhook(...):
    ...
```

Pluggy would then remove ``myhost_oldhook`` from the list. This way all the following work:
- old host - old plugin
- new host - old plugin (with deprecation warning)
- old host - new plugin (because the hook impl is optional)
- new host - new plugin (because pluggy removes the old hook impl from the list and new host only calls the new hook of the plugin)

**Comment by fschulze** [comment]
This PR would make implementing my above proposal for #170 much simpler though.

File: pluggy/hooks.py [hunk 1]
```diff
@@ -202,28 +202,22 @@ def __init__(self, name, hook_execute, specmodule_or_class=None, spec_opts=None)
...
         self.multicall = _multicall
-        self.spec_opts = spec_opts or {}
+        self.spec = None
         if specmodule_or_class is not None:
+            assert spec_opts is not None
             self.set_specification(specmodule_or_class, spec_opts)
 
     def has_spec(self):
-        return self._specmodule_or_class is not None
+        return True if self.spec is not None else False
```
Comment (fschulze) [inline pluggy/hooks.py]: Why not just ``return self.spec is not None``? Much easier to read than the trinary operator.

Comment (tgoodlet) [reply pluggy/hooks.py]: Heh, indeed!

File: pluggy/hooks.py [hunk 2]
```diff
@@ -273,8 +267,10 @@ def __call__(self, *args, **kwargs):
         if args:
             raise TypeError("hook calling supports only keyword arguments")
         assert not self.is_historic()
-        if self.argnames:
-            notincall = set(self.argnames) - set(["__multicall__"]) - set(kwargs.keys())
+        if self.spec:
```
[the remark this answers is not in the document]
Comment (tgoodlet) [reply pluggy/hooks.py]: good call - this looks like a mistake/typo

**Comment by tgoodlet** [comment]
@RonnyPfannschmidt @fschulze changes addressed!

Btw @fschulze we've been having a [discussion on gitter](https://gitter.im/pytest-dev/pluggy?at=5b6b3a4785278d705ea0ebe5).

**Comment by fschulze** [comment]
@tgoodlet the conclusion of that discussion leads to my proposal again to allow deprecation of old hooks without double calls

**Comment by tgoodlet** [comment]
@fschulze yup just keeping you in the loop :)

File: pluggy/manager.py [hunk 3]
```diff
@@ -215,10 +217,10 @@ def _verify_hook(self, hook, hookimpl):
                 "Plugin %r\nhook %r\nhistoric incompatible to hookwrapper"
                 % (hookimpl.plugin_name, hook.name),
             )
-        if hook.warn_on_impl:
-            _warn_for_function(hook.warn_on_impl, hookimpl.function)
+        if hook.spec.warn_on_impl:
```
Comment (tgoodlet) [inline pluggy/manager.py]: I feel like there'll be a bug here if there's no spec.

Comment (tgoodlet) [reply pluggy/manager.py]: Wait, nm. This is never called unless a spec exists.

**Comment by nicoddemus** [comment]
> @nicoddemus do you think we should document this as well or just leave as an internal until more people ask for it?

I think we can keep this an internal detail for now (sorry for the delay)

**Pull request merged**